frame/staking/src/lib.rs

```diff
@@ -924,6 +924,59 @@ pub mod pallet {
 
 		/// Weight information for extrinsics in this pallet.
 		type WeightInfo: WeightInfo;
+
+		/// The list of thresholds separating the various voter bags.
+		///
+		/// Voters are separated into unsorted bags according to their vote weight. This specifies
+		/// the thresholds separating the bags. A voter's bag is the largest bag for which the
+		/// voter's weight is less than or equal to its upper threshold.
+		///
+		/// When voters are iterated, higher bags are iterated completely before lower bags. This
+		/// that iteration is _semi-sorted_: voters of higher weight tend to come before voters of
+		/// lower weight, but peer voters within a particular bag are sorted in insertion order.
+		///
+		/// # Expressing the constant
+		///
+		/// This constant must be sorted in strictly increasing order. Duplicate items are not
+		/// permitted.
+		///
+		/// There is an implied upper limit of `VoteWeight::MAX`; that value does not need to be
+		/// specified within the bag. For any two threshold lists, if one ends with
+		/// `VoteWeight::MAX`, the other one does not, and they are otherwise equal, the two lists
+		/// will behave identically.
+		///
+		/// # Calculation
+		///
+		/// It is recommended to generate the set of thresholds in a geometric series, such that
+		/// there exists some constant ratio such that `threshold[k + 1] == (threshold[k] *
+		/// constant_ratio).max(threshold[k] + 1)` for all `k`.
+		///
+		/// Given the desire to compute `N` bags, the constant ratio can be computed with
+		/// `exp(ln(VoterBags::MAX) / N)`.
+		///
+		/// # Examples
+		///
+		/// - If `VoterBagThresholds::get().is_empty()`, then all voters are put into the same bag,
+		///   and iteration is strictly in insertion order.
+		/// - If `VoterBagThresholds::get().len() == 64`, and the thresholds are determined
+		///   according to the procedure given above, then the constant ratio is equal to 2.
+		/// - If `VoterBagThresholds::get().len() == 200`, and the thresholds are determined
+		///   according to the procedure given above, then the constant ratio is approximately equal
+		///   to 1.248.
+		/// - If the threshold list begins `[1, 2, 3, ...]`, then a voter with weight 0 or 1 will
+		///   fall into bag 0, a voter with weight 2 will fall into bag 1, etc.
+		///
+		/// # Migration
+		///
+		/// In the event that this list ever changes, the `VoterList` data structure will need to be
```

Comment:
How? Do we need to change all of them? If one bag is split into two, will that the managable?

Reply:
As of this iteration of the code, yes, we need to change all of them, and it's not more efficient to split one bag into two. That's because I haven't yet come up with a good indirection method which allows us to go from an actual vote weight to a bag without indexing. 

My best idea so far is to have a function like `upper_threshold_of_containing_bag(v: VoteWeight) -> VoteWeight`, and then store the bags in storage in a map of `VoteWeight -> Bag`. However, I haven't written that yet. If we implement it like this, then on a bag split, we have to touch all nodes contained in the split bag, and on a bag merge, we have to touch all nodes contained in the lower bag.

Reply:
https://github.com/paritytech/substrate/pull/9230/commits/863519f4b8e1710e7c8829af13c9abdd7f272e53 starts on the path of improving that situation, but it's not quite there yet. Currently writing a `fn migrate` which should handle the trickiness of identifying and migrating all appropriate accounts.

Reply:
`fn migrate` ([here](https://github.com/paritytech/substrate/blob/6a8fbeb94c4a7e789a98f604008edd03ee688521/frame/staking/src/voter_bags.rs#L256)) solves that problem.